File: dopamine/packages/core/src/effects/comic.ts
/**
 * Comic Impact (the "BAM! POW!" success effect) as an `EffectFactory`.
 *
 * A hybrid: the jagged starburst + hand-lettered word + ink contours are drawn
 * into ONE offscreen Canvas2D panel each frame; the fragment shader adds the
 * Ben-Day halftone, action lines, flash, noir↔pop styling and casts the light.
 * The program links once in the shared context; the per-fire panel TEXTURE is
 * the instance's own GPU resource (it holds per-frame pixels, so it can't be
 * cached) and is uploaded into each context (light + shadow), then freed on
 * dispose. The conductor owns the overlay / clear / blend / loop.
 */

import { impactScale, impactPresence, IMPACT_MS, IMPACT_HOLD_MS } from "../engine/tempo.js";
import { type ComicRenderParams } from "../engine/mood.js";
import { resolveComicParams } from "../engine/mood.js";
import { shadowGeometry } from "../engine/shadow.js";
import { COMIC_FRAGMENT_SRC, COMIC_VERTEX_SRC } from "../engine/comic-shader.js";
import { drawPanel } from "../engine/comic-renderer.js";
import type { EffectContext, EffectFactory, EffectInstance } from "../framework/effect.js";
import { registerEffect } from "../framework/registry.js";
import type { GLContext } from "../engine/context.js";

const UNIFORMS = [
  "uPanel", "uResolution", "uCenter", "uLife", "uTimeS", "uPresence", "uFlash",
  "uExposure", "uHalftone", "uDotSize", "uSaturation", "uActionLines",
  "uInkBoost", "uSeed", "uStyle", "uC0", "uC1", "uC2",
  "uShadow", "uShadowOffset", "uShadowSoft", "uShadowStrength",
] as const;

function makePanelTexture(glc: GLContext): WebGLTexture {
  const { gl } = glc;
  const tex = gl.createTexture();
  gl.bindTexture(gl.TEXTURE_2D, tex);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
  gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
  return tex!;
}

function createInstance(params: ComicRenderParams, ctx: EffectContext): EffectInstance {
  const [c0, c1, c2] = params.palette;
  const dpr = ctx.dpr;

  // One offscreen Canvas2D panel (word + burst + ink), shared by both passes.
  const panel = document.createElement("canvas");
  const pctx = panel.getContext("2d", { alpha: true })!;

  const lightTex = makePanelTexture(ctx.light);
  const shadowTex = ctx.shadow ? makePanelTexture(ctx.shadow) : null;

  const drawPass = (
    glc: GLContext,
    tex: WebGLTexture,
    life: number,
    elapsedMs: number,
    presence: number,
    flash: number,
    isShadow: boolean,
  ): void => {
    const { gl } = glc;
    const c = glc.canvas;
    const prog = glc.program(COMIC_VERTEX_SRC, COMIC_FRAGMENT_SRC);
    const u = prog.uniforms(UNIFORMS);
    gl.useProgram(prog.program);
    gl.bindVertexArray(glc.vao);

    gl.activeTexture(gl.TEXTURE0);
    gl.bindTexture(gl.TEXTURE_2D, tex);
    gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, true);
    gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, panel);
    gl.uniform1i(u.uPanel, 0);

    gl.uniform2f(u.uResolution, c.width, c.height);
    gl.uniform2f(u.uCenter, c.width * 0.5, c.height * 0.5);
    gl.uniform1f(u.uLife, life);
    gl.uniform1f(u.uTimeS, elapsedMs / 1000);
    gl.uniform1f(u.uPresence, presence);
    gl.uniform1f(u.uFlash, Math.min(flash, 1.2));
    gl.uniform1f(u.uExposure, params.exposure);
    gl.uniform1f(u.uHalftone, params.halftone);
    gl.uniform1f(u.uDotSize, params.dotSize * dpr);
    gl.uniform1f(u.uSaturation, params.saturation);
    gl.uniform1f(u.uActionLines, params.actionLines);
    gl.uniform1f(u.uInkBoost, 1.0 + params.style * 0.4);
    gl.uniform1f(u.uSeed, params.comicSeed);
    gl.uniform1f(u.uStyle, params.style);
    gl.uniform3f(u.uC0, c0.r, c0.g, c0.b);
    gl.uniform3f(u.uC1, c1.r, c1.g, c1.b);
    gl.uniform3f(u.uC2, c2.r, c2.g, c2.b);
    gl.uniform1f(u.uShadow, isShadow ? 1 : 0);
    if (isShadow) {
      const minDim = Math.min(c.width, c.height);
      const sg = shadowGeometry({ minDim, heightFrac: 0.5, amp: presence, style: params.style });
      gl.uniform2f(u.uShadowOffset, sg.offsetX, sg.offsetY);
      gl.uniform1f(u.uShadowSoft, sg.soft);
      gl.uniform1f(u.uShadowStrength, sg.strength);
    }
    gl.drawArrays(gl.TRIANGLES, 0, 3);
  };

  let disposed = false;
  return {
    durationMs: params.durationMs,
    renderAt(elapsedMs: number): void {
      if (disposed) return;
      const c = ctx.light.canvas;
      if (panel.width !== c.width || panel.height !== c.height) {
        panel.width = c.width;
        panel.height = c.height;
      }
      const life = Math.min(Math.max(elapsedMs, 0) / params.durationMs, 1);
      const scale = impactScale(elapsedMs, params.overshoot);
      const presence = impactPresence(life);
      const flash =
        Math.exp(-elapsedMs / (IMPACT_MS * 0.55)) +
        0.25 * Math.exp(-Math.abs(elapsedMs - IMPACT_HOLD_MS * 0.2) / (IMPACT_MS * 0.8));

      // Draw the shared offscreen panel once, then composite into each pass.
      drawPanel(pctx, c.width, c.height, params, scale, presence, dpr);
      if (ctx.shadow && shadowTex) drawPass(ctx.shadow, shadowTex, life, elapsedMs, presence, flash, true);
      drawPass(ctx.light, lightTex, life, elapsedMs, presence, flash, false);
    },
    dispose(): void {
      if (disposed) return;
      disposed = true;
      ctx.light.gl.deleteTexture(lightTex);
      if (ctx.shadow && shadowTex) ctx.shadow.gl.deleteTexture(shadowTex);
    },
  };
}

export const comic: EffectFactory<ComicRenderParams> = {
  name: "comic",
  resolve: (feeling) =>
    resolveComicParams({
      mood: feeling.mood,
      intensity: feeling.intensity,
      whimsy: feeling.whimsy,
      seed: feeling.seed,
    }),
  create: createInstance,
  reducedMotion: { peakMs: 220, holdMs: 360 },
};

export default registerEffect(comic);

File: effects/comic/android/Comic.kt
// Comic Impact as a Dopamine effect on the Android backbone — mirror of the web
// `effect-comic/src/index.ts` + swift's `Comic.swift`.
//
// Per the generalization mandate, the ONLY per-effect code is {the GLSL shader +
// the bespoke tempo + the Canvas panel draw + a tiny config naming its uniforms /
// bindings / frame timing}. Everything else — the `.dope` mapping, the OKLCH
// palette, the registry, the panel runner, the standard uniforms, the shadow
// geometry — is shared backbone. The numeric/palette bag comes verbatim from the
// bundled `.dope` (the SAME bytes as the web), resolved by the shared loader
// (byte-parity proven); the seed-picked SLAMMED word is composed inside the panel
// draw with the shared `pickFromList` (no effect on the numeric/palette parity).
//
// HYBRID: the jagged starburst + hand-lettered word + ink contours are drawn into
// ONE offscreen Canvas panel each frame (ComicPanel.kt); the fragment shader
// (ComicShader.kt) adds the Ben-Day halftone, action lines, flash, noir↔pop
// styling and casts the light. Uses `createPanelInstance` + `PanelConfig` (like
// heartburst), NOT the pure-shader pass runner.

package ai.dopamine.effect.comic

import ai.dopamine.core.DopeDoc
import ai.dopamine.core.DopeResolveInput
import ai.dopamine.core.DopeValue
import ai.dopamine.core.EffectRegistry
import ai.dopamine.core.number
import ai.dopamine.core.parseDope
import ai.dopamine.core.resolveDopeParams
import ai.dopamine.core.resolveTypography
import ai.dopamine.gl.DrawableEffect
import ai.dopamine.gl.EffectContext
import ai.dopamine.gl.EffectInstance
import ai.dopamine.gl.PanelConfig
import ai.dopamine.gl.createPanelInstance
import android.content.Context
import kotlin.math.exp
import kotlin.math.min

class Comic(context: Context) : DrawableEffect {
    override val name: String = "comic"

    // The APK AssetManager — used to load both the `.dope` and (in the panel) the
    // bundled display-face ttf from `assets/fonts/`.
    private val assets = context.assets

    // Load the bundled `.dope` (the EXACT web bytes) from the merged APK assets —
    // proving the data spine is shared verbatim across platforms.
    val doc: DopeDoc = parseDope(
        assets.open("comic.dope.json").bufferedReader(Charsets.UTF_8).use { it.readText() },
    )

    override fun resolve(feeling: DopeResolveInput): Map<String, DopeValue> {
        // Comic declares NO consts; `comicSeed` is the scatter key (the numeric/
        // palette half is byte-identical to the web `resolveDopeParams(DOPE,
        // feeling, {}, "comicSeed")`). The per-mood TYPOGRAPHY (face + curve
        // fields) is composed on top — ADDITIVE, mirroring the web composeComic,
        // so the parity grid stays green.
        val bag = resolveDopeParams(doc, feeling, consts = emptyMap(), scatterKey = "comicSeed").toMutableMap()
        bag.putAll(resolveTypography(doc, feeling.mood, feeling.intensity, feeling.whimsy))
        return bag
    }

    override fun create(params: Map<String, DopeValue>, ctx: EffectContext): EffectInstance =
        createPanelInstance(config, params, ctx)

    // Web factory `reducedMotion: { peakMs: 220, holdMs: 360 }`.
    override val reducedMotionPeakMs: Double = 220.0
    override val reducedMotionHoldMs: Double = 360.0

    // The panel config is per-INSTANCE (not a static val) so its `draw` lambda can
    // close over the AssetManager and load the bundled display faces.
    private val config = PanelConfig(
        vertex = COMIC_VERTEX_SRC,
        fragment = COMIC_FRAGMENT_SRC,
        panelSampler = "uPanel",
        // The shader's non-standard uniforms (the standard half — uCenter,
        // uResolution, uTarget, uLife, uTimeS, uStyle, uAmp, uC0..2, uShadow* —
        // is resolved by the runner). Mirrors the web `index.ts` `uniforms`.
        uniforms = listOf(
            "uPresence", "uFlash", "uExposure", "uHalftone", "uDotSize",
            "uSaturation", "uActionLines", "uInkBoost", "uSeed",
        ),
        // comicSeed drives uSeed; raw seed / overshoot / draw-only geometry
        // (scale, burstPoints, inkWeight) + the dpr-scaled dotSize are NOT
        // auto-bound uniforms (matches the web `bindings`).
        bindings = mapOf(
            "comicSeed" to "uSeed",
            "seed" to null,
            "overshoot" to null,
            "scale" to null,
            "burstPoints" to null,
            "inkWeight" to null,
            "dotSize" to null,
        ),
        // The web `PanelConfig.shadowHeightFrac` is a constant 0.5 (the panel's
        // implied occluder height); no `.dope` param drives it.
        shadowHeightFrac = { 0.5 },
        passUniforms = { _, _, params, density ->
            // dotSize is authored at 1x dpr; multiply by the live density.
            // inkBoost fattens the ink toward the pop end. Both mirror the web
            // `passUniforms` (cross-checked vs swift `packUniforms`).
            mapOf(
                "uDotSize" to (params.number("dotSize") * density).toFloat(),
                "uInkBoost" to (1.0 + params.number("style") * 0.4).toFloat(),
            )
        },
        draw = { canvas, w, h, params, info ->
            // ANIMATED (web parity): redrawn each frame with the LIVE slam scale +
            // presence (recovered from info.elapsedMs / info.life), with the full
            // per-letter typography in the mood-picked bundled face.
            drawComicPanel(
                canvas, assets, params,
                info.elapsedMs, info.life,
                info.centerX, info.centerY,
                info.targetWidthPx, info.targetHeightPx,
                w, h,
            )
        },
        frame = { info, _ ->
            // Mirror the web `frame()`: presence feeds the shadow geometry amp
            // AND uPresence; the impact flash is the fast spike that throws cast
            // light (clamped as before).
            val presence = impactPresence(info.life)
            val flash =
                exp(-info.elapsedMs / (IMPACT_MS * 0.55)) +
                    0.25 * exp(-kotlin.math.abs(info.elapsedMs - IMPACT_HOLD_MS * 0.2) / (IMPACT_MS * 0.8))
            mapOf(
                "amp" to presence.toFloat(),
                "uPresence" to presence.toFloat(),
                "uFlash" to min(flash, 1.2).toFloat(),
            )
        },
    )

    companion object {
        /** Construct + register the effect (needs a Context for the `.dope` asset). */
        fun register(context: Context): Comic {
            val fx = Comic(context.applicationContext)
            EffectRegistry.register(fx)
            return fx
        }
    }
}
